tls/server-tls-threaded.c

```diff
@@ -89,40 +95,39 @@ void* ClientHandler(void* args)
 
     printf("Client %d connected successfully\n", pkg->num);
 
-
     /* Read the client data into our buff array */
     memset(buff, 0, sizeof(buff));
-    if (wolfSSL_read(ssl, buff, sizeof(buff)-1) == -1) {
-        fprintf(stderr, "ERROR: failed to read\n");
-        pkg->open = 1;
-        pthread_exit(NULL);
+    while (wolfSSL_want_read(ssl) || XSTRLEN(buff) == 0) {
+        ret = wolfSSL_read(ssl, buff, sizeof(buff)-1);
+        /* TODO: Currently this thread can get stuck infinitely if client
+         *       disconnects, add timer to abort on a timeout eventually,
+         *       just an example for now so allow for possible stuck condition
+         */
     }
 
     /* Print to stdout any data the client sends */
-    printf("Client %d: %s\n", pkg->num, buff);
+    printf("Client %d said: %s\n", pkg->num, buff);
 
     /* Check for server shutdown command */
     if (strncmp(buff, "shutdown", 8) == 0) {
         printf("Shutdown command issued!\n");
         *pkg->shutdown = 1;
     }
 
-
-
     /* Write our reply into buff */
     memset(buff, 0, sizeof(buff));
     memcpy(buff, reply, strlen(reply));
-    len = strnlen(buff, sizeof(buff));
+    len = XSTRLEN(reply);
 
     /* Reply back to the client */
-    if (wolfSSL_write(ssl, buff, len) != len) {
-        fprintf(stderr, "ERROR: failed to write\n");
-        pkg->open = 1;
-        pthread_exit(NULL);
+    while (wolfSSL_want_write(ssl) || ret < len) {
```

Comment:
do-while loop may fit better and check on return value of ret to see if peer has shut down